### PR TITLE
dynamic unitrim

### PR DESCRIPTION
**Pull request by Rexwang8**
Dynamic Unitrim Generation.

Problem: We currently have to pre-generate the unitrim.json files for a tokenizer. This is used during the unitrim process, which prevents multi-token characters from being cut off between chunks.

This PR implements a dynamic generation and loading of the unitrim data from the encoder.json at start time. This allows for arbitrary tokenizers support for tokenizers (as long as an encoder.json is included). Also includes 1 test case.

*gpt2 unitrim file was found to be slightly incorrect, replaced with updated version.

### CODE REVIEW

File: gpt_bpe.go [hunk 1]
```diff
@@ -938,10 +1082,49 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
 			req = 0
 		} else {
 			req = encoder.unitrim[(*tokens)[tokenIdx]]
+			fmt.Printf("req: %d for token: %d, %s\n", req, tok, encoder.decoder[tok])
+		}
+
+		if !(need+req < 0) {
+			need += req
+		}
+		fmt.Printf("need: %d, req: %d\n", need, req)
```
Comment (wbrown) [inline gpt_bpe.go]: this should probably be removed.

File: gpt_bpe.go [hunk 2]
```diff
@@ -938,10 +1082,49 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
 			req = 0
 		} else {
 			req = encoder.unitrim[(*tokens)[tokenIdx]]
+			fmt.Printf("req: %d for token: %d, %s\n", req, tok, encoder.decoder[tok])
```
Comment (wbrown) [inline gpt_bpe.go]: log.error or something like that for this.

File: gpt_bpe.go [hunk 3]
```diff
@@ -938,10 +1082,49 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
...
+	need := 0
+	//for each token, we add up the number required to complete it
+	for tokenIdx := range *tokens {
+		tok := (*tokens)[tokenIdx]
+		var req int
+		if int(tok) >= len(encoder.unitrim) {
+			// Don't error out on tokens that we don't know about.
+			req = 0
+		} else {
+			//we reference the unitrim array to get the number of tokens required to complete the current token
```
Comment (wbrown) [inline gpt_bpe.go]: preface comments with a space, wrap to 80 columns

File: gpt_bpe.go [hunk 4]
```diff
@@ -938,10 +1082,49 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
...
+			req = encoder.unitrim[(*tokens)[tokenIdx]]
+			fmt.Printf("req: %d for token: %d, %s\n", req, tok, encoder.decoder[tok])
```
Comment (wbrown) [inline gpt_bpe.go]: log rather than print

File: gpt_bpe.go [hunk 5]
```diff
@@ -938,10 +1082,49 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
...
+		}
+
 		if !(need+req < 0) {
 			need += req
 		}
+		fmt.Printf("need: %d, req: %d\n", need, req)
```
Comment (wbrown) [inline gpt_bpe.go]: remove this.

File: gpt_bpe.go [hunk 6]
```diff
@@ -951,6 +1134,7 @@ func (encoder *GPTEncoder) TokensReady(tokens *Tokens) bool {
 			good = tokenIdx + 1
 		}
 	}
+	fmt.Printf("good: %d, len: %d\n", good, len(*tokens))
```
Comment (wbrown) [inline gpt_bpe.go]: remove this.

File: gpt_bpe_test.go [hunk 7]
```diff
@@ -817,15 +871,3 @@ func TestModelDownloadPythiaSharded(t *testing.T) {
 	fmt.Println("All Exists - Looks good.")
 
 }
-
-func TestFunctionForShards(t *testing.T) {
```
Comment (wbrown) [inline gpt_bpe_test.go:821]: why is this deleted?

File: resources/resolver.go [hunk 8]
```diff
@@ -673,6 +674,10 @@ func ResolveVocabId(vocabId string, token string) (*HFConfig, *Resources, error)
 			"json"); unitrim != nil {
 			resources["unitrim.json"] = *unitrim
 		}
+		if encoderJson := GetEmbeddedResource(vocabId + "/encoder." +
+			"json"); encoderJson != nil {
+			resources["encoder.json"] = *encoderJson
+		}
 		if config := GetEmbeddedResource(vocabId + "/encoder." +
```
Comment (wbrown) [inline resources/resolver.go:681]: are we reading `encoder.json` here rather than `config.json`?

**Review by wbrown — CHANGES_REQUESTED** [review]
Some questions, mostly need to remove printfs.

File: gpt_bpe_test.go (hunk 7 above)
Comment (Rexwang8) [reply gpt_bpe_test.go:821]: was a testing function i didn't remove, its incorporated into the TestModelDownloadPythiaSharded test case

File: resources/resolver.go (hunk 8 above)
Comment (Rexwang8) [reply resources/resolver.go:681]: config.json doesn't exist, it appears this was the correct behavior before (I didn't change anything). I'm changing it slightly so it won't try to load it twice and just use the data from the first load.
